allennlp/common/file_utils.py

```diff
@@ -88,42 +91,95 @@ def get_from_cache(url: str, cache_dir: str = None) -> str:
     if cache_dir is None:
         cache_dir = DATASET_CACHE
 
-    os.makedirs(cache_dir, exist_ok=True)
-
     # make HEAD request to check ETag
     response = requests.head(url)
     if response.status_code != 200:
         raise IOError("HEAD request failed for url {}".format(url))
 
     # add ETag to filename if it exists
     etag = response.headers.get("ETag")
-    filename = url_to_filename(url, etag)
-
-    # get cache path to put the file
-    cache_path = os.path.join(cache_dir, filename)
-
-    if not os.path.exists(cache_path):
-        # Download to temporary file, then copy to cache dir once finished.
-        # Otherwise you get corrupt cache entries if the download gets interrupted.
-        _, temp_filename = tempfile.mkstemp()
-        logger.info("%s not found in cache, downloading to %s", url, temp_filename)
-
-        # GET file object
-        req = requests.get(url, stream=True)
-        content_length = req.headers.get('Content-Length')
-        total = int(content_length) if content_length is not None else None
-        progress = Tqdm.tqdm(unit="B", total=total)
-        with open(temp_filename, 'wb') as temp_file:
+
+    with _open_cache_index(cache_dir) as cursor:
+        filename = _url_to_filename(cursor, url, etag)
+
+        # filename will be None if it isn't cached
+        if filename is None:
+            # Download to temporary file, then copy to cache dir once finished.
+            # Otherwise you get corrupt cache entries if the download gets interrupted.
+            temp_file = tempfile.NamedTemporaryFile(delete=False)
```

Comment:
I actually had it as a `with` block but changed it back. I had a few reasons:
1. The `with` block adds a level of nesting (2 levels with the `cache_file` below), but this is just cosmetics
2. The `with` block closes the file, which is nice, but `shutil.copyfile()` occurs before the file is closed, so I need to put a `temp_file.flush()` before the copy or the contents get truncated
3. This way follows the previous logic more closely (e.g., the `logger.info()` call about removing the temp file would just appear at the end of the context manager's scope without an explicit `os.remove()`. I was trying to avoid unnecessary diffs for this first PR in order to highlight what's really different.

These reasons are minor, so I could be easily persuaded to use a `with` block.